internal/tokenclient/tokenclient.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package tokenclient
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/utils/clock"
+
+	"go.pinniped.dev/internal/backoff"
+	"go.pinniped.dev/internal/plog"
+)
+
+type WhatToDoWithTokenFunc func(authenticationv1.TokenRequestStatus, metav1.Duration) error
+
+type TokenClient struct {
+	namespace          string
+	serviceAccountName string
+	k8sClient          kubernetes.Interface
+	whatToDoWithToken  WhatToDoWithTokenFunc
+	expirationSeconds  int64
+	clock              clock.Clock
+	logger             plog.Logger
+}
+
+type Opt func(client *TokenClient)
+
+func WithExpirationSeconds(expirationSeconds int64) Opt {
+	return func(client *TokenClient) {
+		client.expirationSeconds = expirationSeconds
+	}
+}
+
+func New(
+	namespace string,
+	serviceAccountName string,
+	k8sClient kubernetes.Interface,
+	whatToDoWithToken WhatToDoWithTokenFunc,
+	logger plog.Logger,
+	opts ...Opt,
+) *TokenClient {
+	client := &TokenClient{
+		namespace:          namespace,
+		serviceAccountName: serviceAccountName,
+		k8sClient:          k8sClient,
+		whatToDoWithToken:  whatToDoWithToken,
+		expirationSeconds:  600,
+		clock:              clock.RealClock{},
+		logger:             logger,
+	}
+	for _, opt := range opts {
+		opt(client)
+	}
+	return client
+}
+
+type howToFetchTokenFromAPIServer func(tokenRequest *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error)
+
+func (tokenClient TokenClient) Start(ctx context.Context) {
+	sleeper := make(chan time.Time, 1)
+	// Make sure that the <-sleeper below gets run once immediately.
+	sleeper <- time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			tokenClient.logger.Info("TokenClient was cancelled and is stopping")
+			return
+		case <-sleeper:
```

Comment:
Looks like it would be easy to extract a function for all the code in this `case`.